File: .grepa/scripts/grepa-list.py
#!/usr/bin/env python3

# :ga:meta grepa-list script - discovers all grep-anchors in your codebase

import json
import subprocess
import sys
import re
import argparse
from datetime import datetime
from pathlib import Path
from collections import defaultdict
from typing import Dict, List, Tuple, Any

# Configuration
DEFAULT_ANCHOR = ':ga:'
OUTPUT_FILE = Path.cwd() / '.grepa' / 'grepa-list.json'


def is_in_code_block(filename: str, line_num: int, anchor_match: str) -> bool:
    """Check if a match is inside a code block or code fence."""
    # Only check markdown files
    if not filename.endswith(('.md', '.markdown', '.mdx')):
        return False
    
    try:
        with open(filename, 'r') as f:
            lines = f.readlines()
        
        # Check if we're inside a code fence
        in_code_fence = False
        for i, line in enumerate(lines[:line_num], 1):
            if line.strip().startswith('```'):
                in_code_fence = not in_code_fence
        
        # If we're in a code fence, it's an example
        if in_code_fence:
            return True
        
        # Check if the line itself is indented (code block)
        if line_num <= len(lines):
            line = lines[line_num - 1]
            # Check for 4-space indentation (markdown code block)
            if line.startswith('    ') and anchor_match in line:
                return True
        
        return False
    except:
        # If we can't read the file, assume it's not in a code block
        return False


def find_grep_anchors(anchor: str = DEFAULT_ANCHOR, ignore_md: bool = False, ignore_examples: bool = False) -> Dict[str, Any]:
    """Find all grep-anchors in the codebase using ripgrep."""
    try:
        # Run ripgrep to find all anchors
        # -n: line numbers, -o: only matching, --no-heading: compact format
        cmd = ['rg', '-n', '-o', f'{anchor}[^\\s]*', '--no-heading']
        
        # Add glob patterns to ignore markdown files if requested
        if ignore_md:
            cmd.extend(['-g', '!*.md', '-g', '!*.markdown', '-g', '!*.mdx'])
        
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        if result.returncode == 1:
            # No matches found
            return {'tags': {}, 'files': {}, 'totalAnchors': 0, 'anchor': anchor}
        elif result.returncode != 0:
            # Other error
            raise subprocess.CalledProcessError(result.returncode, cmd, result.stderr)
        
        output = result.stdout.strip()
        if not output:
            return {'tags': {}, 'files': {}, 'totalAnchors': 0, 'anchor': anchor}
        
        lines = output.split('\n')
        tags = defaultdict(lambda: {
            'count': 0,
            'files': defaultdict(list),
            'firstSeen': None
        })
        files = defaultdict(lambda: {
            'totalAnchors': 0,
            'tags': defaultdict(int),
            'lines': []
        })
        
        # Compile regex for tag extraction
        tag_pattern = re.compile(f'{re.escape(anchor)}([^\\s,\\]\\}}]+)')
        
        for line in lines:
            # Parse format: filename:line:match
            match = re.match(r'^(.+?):(\d+):(.+)$', line)
            if not match:
                continue
            
            filename, line_num, full_match = match.groups()
            line_num = int(line_num)
            
            # Skip if we're ignoring examples and this is in a code block
            if ignore_examples and is_in_code_block(filename, line_num, full_match):
                continue
            
            # Extract the tag from the full match
            tag_match = tag_pattern.match(full_match)
            if not tag_match:
                continue
            
            tag = tag_match.group(1)
            
            # Update tag statistics
            tags[tag]['count'] += 1
            tags[tag]['files'][filename].append(line_num)
            
            if tags[tag]['firstSeen'] is None:
                tags[tag]['firstSeen'] = {'file': filename, 'line': line_num}
            
            # Update file statistics
            files[filename]['totalAnchors'] += 1
            files[filename]['tags'][tag] += 1
            files[filename]['lines'].append(line_num)
        
        # Convert defaultdicts to regular dicts and sort lines
        tags_dict = {}
        for tag, data in tags.items():
            tags_dict[tag] = {
                'count': data['count'],
                'files': {f: sorted(lines) for f, lines in data['files'].items()},
                'firstSeen': data['firstSeen']
            }
        
        files_dict = {}
        for filename, data in files.items():
            files_dict[filename] = {
                'totalAnchors': data['totalAnchors'],
                'tags': dict(data['tags']),
                'lines': sorted(data['lines'])
            }
        
        return {
            'tags': tags_dict,
            'files': files_dict,
            'totalAnchors': len(lines),
            'anchor': anchor
        }
        
    except FileNotFoundError:
        print('❌ Error: ripgrep (rg) is not installed or not in PATH')
        print('   Install it from: https://github.com/BurntSushi/ripgrep')
        sys.exit(1)
    except subprocess.CalledProcessError as e:
        print(f'❌ Error running ripgrep: {e.stderr}')
        sys.exit(1)


def generate_report(data: Dict[str, Any]) -> Dict[str, Any]:
    """Generate a comprehensive report from the anchor data."""
    tags = data['tags']
    files = data['files']
    total_anchors = data['totalAnchors']
    anchor = data['anchor']
    
    # Sort tags by count (descending)
    sorted_tags = sorted(tags.items(), key=lambda x: x[1]['count'], reverse=True)
    
    # Sort files by anchor count (descending)
    sorted_files = sorted(files.items(), key=lambda x: x[1]['totalAnchors'], reverse=True)
    
    report = {
        '_comment': ':ga:meta Generated grep-anchor inventory - DO NOT EDIT MANUALLY',
        'generated': datetime.now().isoformat(),
        'anchor': anchor,
        'summary': {
            'totalAnchors': total_anchors,
            'uniqueTags': len(tags),
            'filesWithAnchors': len(files)
        },
        'tags': dict(sorted_tags),
        'files': dict(sorted_files),
        'topTags': [
            {
                'tag': tag,
                'count': data['count'],
                'fileCount': len(data['files'])
            }
            for tag, data in sorted_tags[:10]
        ],
        'topFiles': [
            {
                'file': filename,
                'anchors': data['totalAnchors'],
                'uniqueTags': len(data['tags'])
            }
            for filename, data in sorted_files[:10]
        ]
    }
    
    return report


def main():
    """Main execution function."""
    # Parse command line arguments
    parser = argparse.ArgumentParser(description='Discover all grep-anchors in your codebase')
    parser.add_argument('anchor', nargs='?', default=DEFAULT_ANCHOR, 
                        help=f'Custom anchor pattern (default: {DEFAULT_ANCHOR})')
    parser.add_argument('--ignore-md', action='store_true',
                        help='Ignore markdown files (*.md, *.markdown, *.mdx)')
    parser.add_argument('--ignore-examples', action='store_true',
                        help='Ignore grep-anchors inside code blocks and code fences')
    
    args = parser.parse_args()
    
    print('🔍 Searching for grep-anchors...')
    
    data = find_grep_anchors(args.anchor, args.ignore_md, args.ignore_examples)
    report = generate_report(data)
    
    # Ensure .grepa directory exists
    OUTPUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    
    # Write JSONC file with pretty printing
    with open(OUTPUT_FILE, 'w') as f:
        json.dump(report, f, indent=2)
    
    # Print summary
    print(f'\n✅ Grep-anchor inventory generated!')
    print(f'📍 Anchor pattern: {anchor}')
    print(f'📊 Found {report["summary"]["totalAnchors"]} anchors across {report["summary"]["filesWithAnchors"]} files')
    print(f'🏷️  Discovered {report["summary"]["uniqueTags"]} unique tags')
    print(f'📄 Report saved to: {OUTPUT_FILE}')
    
    if report['topTags']:
        print('\n🔝 Top 5 tags:')
        for item in report['topTags'][:5]:
            print(f'   {item["tag"]}: {item["count"]} uses in {item["fileCount"]} file(s)')


if __name__ == '__main__':
    main() 

File: .grepa/scripts/grepa-list.js
#!/usr/bin/env node

// :ga:meta grepa-list script - discovers all grep-anchors in your codebase

import { execSync } from 'child_process';
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { argv, exit } from 'process';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Configuration
const DEFAULT_ANCHOR = ':ga:';
const CONFIG_FILE = join(process.cwd(), '.grepa', 'grepa-list.config.json');
const DEFAULT_OUTPUT_FILE = join(process.cwd(), '.grepa', 'grepa-list.json');

// Precompiled regex patterns for performance
const INLINE_CODE_PATTERN = /(?<!`)`(?!``)[^`]+`(?!`)/g;
const LINK_PATTERN = /\[([^\]]+)\]\([^\)]+\)/g;
const BLOCKQUOTE_PATTERN = /^\s*>/;
const HEADING_PATTERN = /^#{1,6}\s+/;

// Load configuration
function loadConfig() {
  if (existsSync(CONFIG_FILE)) {
    try {
      const content = readFileSync(CONFIG_FILE, 'utf8');
      return JSON.parse(content);
    } catch (e) {
      console.warn(`⚠️  Warning: Could not load config file: ${e.message}`);
    }
  }
  return {};
}

// Parse command line arguments
function parseArgs(config) {
  const args = {
    anchor: config.anchor || DEFAULT_ANCHOR,
    ignorePatterns: [],
    ignoreExamples: false,
    noGitignore: false,
    configPath: null
  };
  
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === '--ignore' && i + 1 < argv.length) {
      args.ignorePatterns.push(argv[++i]);
    } else if (argv[i] === '--ignore-examples') {
      args.ignoreExamples = true;
    } else if (argv[i] === '--no-gitignore') {
      args.noGitignore = true;
    } else if (argv[i] === '--config' && i + 1 < argv.length) {
      args.configPath = argv[++i];
    } else if (!argv[i].startsWith('--')) {
      args.anchor = argv[i];
    }
  }
  
  return args;
}

// Resolve ignore patterns from aliases and custom patterns
function resolveIgnorePatterns(patterns, config) {
  const globs = [];
  const ignoreConfig = config.ignore || {};
  const patternDefs = ignoreConfig.patterns || {};
  
  for (const pattern of patterns) {
    if (patternDefs[pattern]) {
      // It's a predefined pattern alias
      globs.push(...(patternDefs[pattern].globs || []));
    } else {
      // It's a custom pattern
      globs.push(pattern);
    }
  }
  
  return globs;
}

function isInMarkdownConstruct(filename, lineNum, anchorMatch, config, fileCache = null) {
  // Only check markdown files
  if (!filename.match(/\.(md|markdown|mdx)$/)) {
    return false;
  }
  
  try {
    // Use cached file contents if available
    let lines;
    if (fileCache && fileCache[filename]) {
      lines = fileCache[filename];
    } else {
      const content = readFileSync(filename, 'utf8');
      lines = content.split('\n');
      if (fileCache) {
        fileCache[filename] = lines;
      }
    }
    
    if (lineNum > lines.length) {
      return false;
    }
    
    const currentLine = lines[lineNum - 1];
    const examplesConfig = config.examples || {};
    
    // Check code fences (```)
    if (examplesConfig.detectCodeFences !== false) {
      let inCodeFence = false;
      for (let i = 0; i < lineNum - 1; i++) {
        if (lines[i].trim().startsWith('```')) {
          inCodeFence = !inCodeFence;
        }
      }
      if (inCodeFence) {
        return true;
      }
    }
    
    // Check code blocks (4-space indentation)
    if (examplesConfig.detectCodeBlocks !== false) {
      if (currentLine.startsWith('    ') && currentLine.includes(anchorMatch)) {
        return true;
      }
    }
    
    // Check headings (lines starting with #)
    if (examplesConfig.detectHeadings) {
      if (HEADING_PATTERN.test(currentLine)) {
        return true;
      }
    }
    
    // Check inline code (text between backticks)
    if (examplesConfig.detectInlineCode) {
      const anchorPos = currentLine.indexOf(anchorMatch);
      if (anchorPos !== -1) {
        const matches = [...currentLine.matchAll(INLINE_CODE_PATTERN)];
        for (const match of matches) {
          if (match.index <= anchorPos && anchorPos <= match.index + match[0].length) {
            return true;
          }
        }
      }
    }
    
    // Check links [text](url)
    if (examplesConfig.detectLinks) {
      const anchorPos = currentLine.indexOf(anchorMatch);
      if (anchorPos !== -1) {
        const matches = [...currentLine.matchAll(LINK_PATTERN)];
        for (const match of matches) {
          if (match.index <= anchorPos && anchorPos <= match.index + match[0].length) {
            return true;
          }
        }
      }
    }
    
    // Check blockquotes (lines starting with >)
    if (examplesConfig.detectBlockquotes) {
      if (BLOCKQUOTE_PATTERN.test(currentLine)) {
        return true;
      }
    }
    
    // Check HTML comments
    if (examplesConfig.detectHtmlComments) {
      // Check if we're inside an HTML comment
      const fullText = lines.slice(0, lineNum).join('\n');
      const commentCount = (fullText.match(/<!--/g) || []).length;
      const commentEndCount = (fullText.match(/-->/g) || []).length;
      if (commentCount > commentEndCount) {
        return true;
      }
      // Also check if the current line has a comment
      if (currentLine.includes('<!--') && currentLine.includes(anchorMatch)) {
        const commentStart = currentLine.indexOf('<!--');
        const anchorPos = currentLine.indexOf(anchorMatch);
        const commentEnd = currentLine.indexOf('-->', commentStart);
        if (commentEnd === -1) {  // Comment continues on next line
          if (anchorPos > commentStart) {
            return true;
          }
        } else if (commentStart < anchorPos && anchorPos < commentEnd) {
          return true;
        }
      }
    }
    
    // Check front matter (YAML/TOML between --- or +++)
    if (examplesConfig.detectFrontMatter) {
      if (lineNum === 1) {
        return false;
      }
      // Check if we're in front matter
      if (lines[0].trim() === '---' || lines[0].trim() === '+++') {
        const delimiter = lines[0].trim();
        for (let i = 1; i < lineNum - 1; i++) {
          if (lines[i].trim() === delimiter) {
            return false;  // We passed the end of front matter
          }
        }
        return true;  // Still in front matter
      }
    }
    
    return false;
  } catch {
    // If we can't read the file, assume it's not in a special construct
    return false;
  }
}

function findGrepAnchors(anchor = DEFAULT_ANCHOR, ignorePatterns = [], ignoreExamples = false, config = {}) {
  try {
    // Create file cache for markdown construct detection
    const fileCache = {};
    
    // Run ripgrep to find all anchors
    // -n: line numbers, -o: only matching, --no-heading: compact format
    let rgCommand = `rg -n -o "${anchor}[^\\s]*" --no-heading`;
    
    // Check if we should respect gitignore
    const ignoreConfig = config.ignore || {};
    if (ignoreConfig.respectGitignore) {
      // Check if .gitignore exists
      const gitignorePath = join(process.cwd(), '.gitignore');
      if (existsSync(gitignorePath)) {
        // Add .gitignore patterns
        rgCommand += ` --ignore-file "${gitignorePath}"`;
      }
      // Otherwise ripgrep respects .gitignore by default
    } else {
      // Explicitly disable gitignore
      rgCommand += ' --no-ignore-vcs';
    }
    
    // Resolve ignore patterns from config and command line
    const allPatterns = [];
    
    // Add active patterns from config file
    const patternDefs = ignoreConfig.patterns || {};
    for (const [patternName, patternDef] of Object.entries(patternDefs)) {
      if (patternDef.active) {
        allPatterns.push(patternName);
      }
    }
    
    // Add custom patterns
    allPatterns.push(...(ignoreConfig.custom || []));
    
    // Add patterns from command line (override config)
    allPatterns.push(...ignorePatterns);
    
    // Resolve all patterns to globs
    if (allPatterns.length > 0) {
      const globs = resolveIgnorePatterns(allPatterns, config);
      for (const glob of globs) {
        rgCommand += ` -g "!${glob}"`;
      }
    }
    
    const output = execSync(rgCommand, { encoding: 'utf8', stdio: ['pipe', 'pipe', 'ignore'] });
    
    if (!output.trim()) {
      return { tags: {}, files: {}, totalAnchors: 0 };
    }
    
    const lines = output.trim().split('\n');
    const tags = {};
    const files = {};
    
    lines.forEach(line => {
      // Parse format: filename:line:match
      const match = line.match(/^(.+?):(\d+):(.+)$/);
      if (!match) return;
      
      const [, filename, lineNum, fullMatch] = match;
      
      // Skip if we're ignoring examples and this is in a markdown construct
      if (ignoreExamples && isInMarkdownConstruct(filename, parseInt(lineNum), fullMatch, config, fileCache)) {
        return;
      }
      
      // Extract the tag from the full match
      const tagMatch = fullMatch.match(new RegExp(`${anchor}([^\\s,\\]\\}]+)`));
      if (!tagMatch) return;
      
      const tag = tagMatch[1];
      
      // Update tag statistics
      if (!tags[tag]) {
        tags[tag] = {
          count: 0,
          files: {},
          firstSeen: { file: filename, line: parseInt(lineNum) }
        };
      }
      
      tags[tag].count++;
      
      // Track files for this tag
      if (!tags[tag].files[filename]) {
        tags[tag].files[filename] = [];
      }
      tags[tag].files[filename].push(parseInt(lineNum));
      
      // Update file statistics
      if (!files[filename]) {
        files[filename] = {
          totalAnchors: 0,
          tags: {},
          lines: []
        };
      }
      
      files[filename].totalAnchors++;
      files[filename].tags[tag] = (files[filename].tags[tag] || 0) + 1;
      files[filename].lines.push(parseInt(lineNum));
    });
    
    // Sort lines in each file
    Object.values(files).forEach(file => {
      file.lines.sort((a, b) => a - b);
    });
    
    // Sort lines in each tag's files
    Object.values(tags).forEach(tag => {
      Object.values(tag.files).forEach(lines => {
        lines.sort((a, b) => a - b);
      });
    });
    
    return {
      tags,
      files,
      totalAnchors: lines.length,
      anchor
    };
    
  } catch (error) {
    if (error.status === 1) {
      // No matches found
      return { tags: {}, files: {}, totalAnchors: 0, anchor };
    }
    throw error;
  }
}

function generateReport(data) {
  const { tags, files, totalAnchors, anchor } = data;
  
  // Sort tags by count (descending)
  const sortedTags = Object.entries(tags)
    .sort(([, a], [, b]) => b.count - a.count);
  
  // Sort files by anchor count (descending)
  const sortedFiles = Object.entries(files)
    .sort(([, a], [, b]) => b.totalAnchors - a.totalAnchors);
  
  const report = {
    _comment: ":ga:meta Generated grep-anchor inventory - DO NOT EDIT MANUALLY",
    generated: new Date().toISOString(),
    anchor: anchor,
    summary: {
      totalAnchors,
      uniqueTags: Object.keys(tags).length,
      filesWithAnchors: Object.keys(files).length
    },
    tags: Object.fromEntries(sortedTags),
    files: Object.fromEntries(sortedFiles),
    topTags: sortedTags.slice(0, 10).map(([tag, data]) => ({
      tag,
      count: data.count,
      fileCount: Object.keys(data.files).length
    })),
    topFiles: sortedFiles.slice(0, 10).map(([file, data]) => ({
      file,
      anchors: data.totalAnchors,
      uniqueTags: Object.keys(data.tags).length
    }))
  };
  
  return report;
}

// Main execution
function main() {
  console.log('🔍 Searching for grep-anchors...');
  
  try {
    // Load configuration
    let config = loadConfig();
    
    // Parse command line arguments
    const args = parseArgs(config);
    
    // Load custom config if specified
    if (args.configPath) {
      if (existsSync(args.configPath)) {
        try {
          const content = readFileSync(args.configPath, 'utf8');
          config = JSON.parse(content);
        } catch (e) {
          console.error(`❌ Error loading custom config: ${e.message}`);
          exit(1);
        }
      }
    }
    
    // Override config with command line args
    const ignoreExamples = args.ignoreExamples || (config.examples?.ignore || false);
    
    // Handle gitignore override
    if (args.noGitignore) {
      if (!config.ignore) config.ignore = {};
      config.ignore.respectGitignore = false;
    }
    
    const data = findGrepAnchors(args.anchor, args.ignorePatterns, ignoreExamples, config);
    const report = generateReport(data);
    
    // Get output file from config or use default
    const outputConfig = config.output || {};
    const outputFile = join(process.cwd(), outputConfig.file || '.grepa/inventory.generated.json');
    const indent = outputConfig.indent || 2;
    
    // Ensure output directory exists
    const outputDir = dirname(outputFile);
    execSync(`mkdir -p "${outputDir}"`, { stdio: 'ignore' });
    
    // Write JSONC file with pretty printing
    const jsonContent = JSON.stringify(report, null, indent);
    writeFileSync(outputFile, jsonContent);
    
    // Gather all active patterns for display
    const allActivePatterns = [];
    const ignoreConfig = config.ignore || {};
    const patternDefs = ignoreConfig.patterns || {};
    
    // Add active named patterns
    for (const [patternName, patternDef] of Object.entries(patternDefs)) {
      if (patternDef.active) {
        allActivePatterns.push(patternName);
      }
    }
    
    // Add custom patterns
    allActivePatterns.push(...(ignoreConfig.custom || []));
    allActivePatterns.push(...args.ignorePatterns);
    
    // Print summary
    const displayConfig = config.display || {};
    if (displayConfig.showSummary !== false) {
      console.log(`\n✅ Grep-anchor inventory generated!`);
      console.log(`📍 Anchor pattern: ${args.anchor}`);
      if (config.ignore?.respectGitignore) {
        console.log(`📁 Respecting .gitignore patterns`);
      }
      if (allActivePatterns.length > 0) {
        console.log(`🚫 Ignored patterns: ${allActivePatterns.join(', ')}`);
      }
      if (ignoreExamples) {
        console.log(`🚫 Ignored code examples`);
      }
      console.log(`📊 Found ${report.summary.totalAnchors} anchors across ${report.summary.filesWithAnchors} files`);
      console.log(`🏷️  Discovered ${report.summary.uniqueTags} unique tags`);
      console.log(`📄 Inventory saved to: ${outputFile}`);
    
      if (report.topTags.length > 0) {
        const topCount = displayConfig.topTagsCount || 5;
        console.log(`\n🔝 Top ${topCount} tags:`);
        report.topTags.slice(0, topCount).forEach(({ tag, count, fileCount }) => {
          console.log(`   ${tag}: ${count} uses in ${fileCount} file(s)`);
        });
      }
    }
    
  } catch (error) {
    if (error.code === 'ENOENT' && error.message.includes('rg')) {
      console.error('❌ Error: ripgrep (rg) is not installed or not in PATH');
      console.error('   Install it from: https://github.com/BurntSushi/ripgrep');
      process.exit(1);
    }
    
    console.error('❌ Error:', error.message);
    process.exit(1);
  }
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
} 